src/vec/drain.rs

```diff
@@ -0,0 +1,163 @@
+use core::{
+    fmt,
+    iter::FusedIterator,
+    mem::{self, size_of},
+    ptr::{self, NonNull},
+    slice,
+};
+
+use super::Vec;
+
+/// A draining iterator for `Vec<T, N>`.
+///
+/// This `struct` is created by [`Vec::drain`].
+/// See its documentation for more.
+///
+/// # Example
+///
+/// ```
+/// use heapless::{vec, Vec};
+///
+/// let mut v = Vec::from_array([0, 1, 2]);
+/// let iter: vec::Drain<'_, _, 3> = v.drain(..);
+/// ```
+pub struct Drain<'a, T: 'a, const N: usize> {
+    /// Index of tail to preserve
+    pub(super) tail_start: usize,
+    /// Length of tail
+    pub(super) tail_len: usize,
+    /// Current remaining range to remove
+    pub(super) iter: slice::Iter<'a, T>,
+    pub(super) vec: NonNull<Vec<T, N>>,
```

Comment:
If this was built on top of #425 this would allow removing the `const N: usize` generic, and replace the pointer with `NonNull<VecView<T>>`. #425 does not offer the `View` functionality for `String` to keep the PR small, but it would be easy to add it.

It is worth considering because merging this PR and releasing before merging #425 would make it a breaking change to remove  the const generic, so there would need to be 2 implementation of `drain`, one for the view and one for the owned type.

Reply:
Yes, I agree it makes sense to base this on `VecView` instead.